proto/errors.go

```diff
@@ -73,16 +73,30 @@ var (
 type KafkaError struct {
 	errno   int16
 	message string
+	// An additional message explaining cases where the client returns this error internally
+	// instead of directly from a Kafka broker.
+	syntheticMessage string
 }
 
 func (err *KafkaError) Error() string {
+	if err.syntheticMessage != "" {
+		fmt.Sprintf("%s (%d %s)", err.message, err.errno, err.syntheticMessage)
+	}
 	return fmt.Sprintf("%s (%d)", err.message, err.errno)
 }
 
 func (err *KafkaError) Errno() int {
 	return int(err.errno)
 }
 
+func (err *KafkaError) SyntheticError() string {
+	return err.syntheticMessage
+}
+
+func (err *KafkaError) NewSynthetic(message string) *KafkaError {
+	return &KafkaError{err.errno, err.message, message}
```

Comment:
Okay, so I think the biggest fear I have is changing something and having those 'if err == xyz' conditions failing. That's pretty catastrophic to behavior if we, say, stop doing metadata refreshes when we should or something.

I kind of like the relabelled Int approach but understand that's a lot of work. Is there any way with the first approach to make sure we don't end up with invisible error cases?